### DISCUSSION TITLE
Moddable Package Manager

### DISCUSSION THREAD

**Discussion by dashcraft** [Moddable-OpenSource/moddable#654] [Ideas]
Thoughts on a moddable package manager?

Having a centralized repository for moddable compatible code would be nice, as well as a way
of managing package installations and inclusions for moddable projects. Since there is a predefined
manifest.json as well as declared includes, there already exists infrastructure to support this kind
of feature/new project.

I'll also volunteer to help out on the project if nodejs/typescript/js is used.

**Comment by phoddie** [comment]
@rwaldron - I recall you had some thoughts on this.

And we should call it mpm. That won't be confusing at all.

**Comment by wilberforce** [comment]
My "solution" is to have a common folder at the base: `/core/modules`
![image](https://user-images.githubusercontent.com/5050824/121280091-cd4ab700-c929-11eb-8b9a-9f5c9eeec531.png)

Then in the actual project I refer to the common code  the `manifest.js`

```
"modules": {
		"*": [
			"./main",
			"../core/utils",
			"./www/src/store",
			"./display",
			"$(MODULES)/drivers/onewire/*",
			"../M5c-tilt/tilt",
			"../core/modules/loadmod",
			"../core/modules/pid",
			"../core/modules/resServer",
			"../core/modules/httpSend",
			"../core/modules/modelServer",
```

This allows files common across projects to be in their own git repository.

Node package managers nests dependancies whereas in moddable - everthing is flat. 
 @dashcraft Are you proposing versioning?

**Reply by dashcraft** [reply]
- I like the idea of a root level folder, i'd be interested in other naming conventions fo the folder

- For versioning, i think it will be necessary in the future but I'm not too keen on solving diverging dependency problems and other dependency/module resolutions when there are conflicts.

It's possible we could keep a really flat dependency tree and manage versions similar to python or rust?
( say python or rust while being mostly a js dev, because npm/yarn has been a fiasco from get go )

**Reply by dashcraft** [reply]
I'm also keen on this package manager also being the starting point for new projects.
Mcconfig, mccrun, xsbug etc are all good command line tools, I think another one would
be for scaffolding out simple setups for typescript, js, or etc.

I.e. npm init scaffolds out projects and handles dependencies, prisma-cli more or less, expo-cli, etc

**Comment by tve** [comment]
> Having a centralized repository for moddable compatible code would be nice

Is there such a repo at this point?

**Reply by dashcraft** [reply]
Only the moddable source code/packages themselves

**Comment by tve** [comment]
I have to admit I'm mostly interested in a directory, i.e. a way to find modules published by others. I'm most likely to hack them up anyway so I'm not looking for download & installation automation as much. I know this may not be the popular opinion, though. E.g., an "awesome-moddable" repo would already be a good start for me.

**Reply by phoddie** [reply]
FWIW – we did some work recently to allow [Git repos to be referenced directly from a manifest](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/manifest.md#include-git). When you build the project with `mcconfig` or `mcrun`, the repos are automatically cloned. It isn't intended to replace a package manager. It is a lightweight, decentralized way to bring-in modules without having to manually download, update paths, etc. Perhaps paired with a directory and a little support in xs-dev it is a very practical tool.
